### ISSUE TITLE
Unable to run on GPU using Intel drivers

### ISSUE THREAD

**Issue by tom91136** [codeplaysoftware/computecpp-sdk#210]
I was unable to run my sycl program on a supported configuration:
CPU: Core Intel i7 6770HQ(Iris Pro 580) and Intel i7 8850H(UHD 630)
OS: CentOS 7
Drivers:  Intel CPU Runtime for OpenCL 18.1
computecpp_info output:
```
********************************************************************************

ComputeCpp Info (CE 1.1.3)

SYCL 1.2.1 revision 3

********************************************************************************

Toolchain information:

GLIBC version: 2.17
GLIBCXX: 20150623
This version of libstdc++ is supported.

********************************************************************************


Device Info:

Discovered 1 devices matching:
  platform    : <any>
  device type : <any>

--------------------------------------------------------------------------------
Device 0:

  Device is supported                     : UNTESTED - Device not tested on this OS
  CL_DEVICE_NAME                          : Intel(R) Gen9 HD Graphics NEO
  CL_DEVICE_VENDOR                        : Intel(R) Corporation
  CL_DRIVER_VERSION                       : 19.29.13530
  CL_DEVICE_TYPE                          : CL_DEVICE_TYPE_GPU 

If you encounter problems when using any of these OpenCL devices, please consult
this website for known issues:
https://computecpp.codeplay.com/releases/v1.1.3/platform-support-notes

********************************************************************************
```

The program crashes on first instance of kernel invocation:

```
Error: [ComputeCpp:RT0100] Failed to build program (<Build log for program 0x325c980 device 0 (size 1375280):

error: undefined reference to `__assert_fail()'

error: backend compiler failed build.

1:1:3: error: source file is not valid UTF-8
BC<C0><DE>!<U+000C>
  ^
1:1:4: error: source file is not valid UTF-8
BC<C0><DE>!<U+000C>
      ^
1:1:1: error: unknown type name 'BC'
BC<C0><DE>!<U+000C>
^
1:1:5: error: expected identifier or '('
BC<C0><DE>!<U+000C>
          ^
1:10:2: error: source file is not valid UTF-8
<U+0007><81>#<91>A<C8><U+0004>I<U+0006><U+0010>29<92><U+0001><84><U+000C>%<U+0005><U+0008><U+0019><U+001E><U+0004><8B>b<80><U+0014>E<U+0002>B<92><U+000B>B<A4><U+0010>2<U+0014>8<U+0008><U+0018>I
        ^
1:10:4: error: source file is not valid UTF-8
<U+0007><81>#<91>A<C8><U+0004>I<U+0006><U+0010>29<92><U+0001><84><U+000C>%<U+0005><U+0008><U+0019><U+001E><U+0004><8B>b<80><U+0014>E<U+0002>B<92><U+000B>B<A4><U+0010>2<U+0014>8<U+0008><U+0018>I
             ^
1:10:6: error: source file is not valid UTF-8
<U+0007><81>#<91>A<C8><U+0004>I<U+0006><U+0010>29<92><U+0001><84><U+000C>%<U+0005><U+0008><U+0019><U+001E><U+0004><8B>b<80><U+0014>E<U+0002>B<92><U+000B>B<A4><U+0010>2<U+0014>8<U+0008><U+0018>I
                  ^
1:10:13: error: source file is not valid UTF-8
<U+0007><81>#<91>A<C8><U+0004>I<U+0006><U+0010>29<92><U+0001><84><U+000C>%<U+0005><U+0008><U+0019><U+001E><U+0004><8B>b<80><U+0014>E<U+0002>B<92><U+000B>B<A4><U+0010>2<U+0014>8<U+0008><U+0018>I
                                                 ^
1:10:15: error: source file is not valid UTF-8
<U+0007><81>#<91>A<C8><U+0004>I<U+0006><U+0010>29<92><U+0001><84><U+000C>%<U+0005><U+0008><U+0019><U+001E><U+0004><8B>b<80><U+0014>E<U+0002>B<92><U+000B>B<A4><U+0010>2<U+0014>8<U+0008><U+0018>I
                                                             ^
1:10:23: error: source file is not valid UTF-8
<U+0007><81>#<91>A<C8><U+0004>I<U+0006><U+0010>29<92><U+0001><84><U+000C>%<U+0005><U+0008><U+0019><U+001E><U+0004><8B>b<80><U+0014>E<U+0002>B<92><U+000B>B<A4><U+0010>2<U+0014>8<U+0008><U+0018>I
                                                                                                                  ^
1:10:25: error: source file is not valid UTF-8
<U+0007><81>#<91>A<C8><U+0004>I<U+0006><U+0010>29<92><U+0001><84><U+000C>%<U+0005><U+0008><U+0019><U+001E><U+0004><8B>b<80><U+0014>E<U+0002>B<92><U+000B>B<A4><U+0010>2<U+0014>8<U+0008><U+0018>I
                                                                                                                       ^
1:10:30: error: source file is not valid UTF-8
<U+0007><81>#<91>A<C8><U+0004>I<U+0006><U+0010>29<92><U+0001><84><U+000C>%<U+0005><U+0008><U+0019><U+001E><U+0004><8B>b<80><U+0014>E<U+0002>B<92><U+000B>B<A4><U+0010>2<U+0014>8<U+0008><U+0018>I
                                                                                                                                             ^
1:10:33: error: source file is not valid UTF-8
<U+0007><81>#<91>A<C8><U+0004>I<U+0006><U+0010>29<92><U+0001><84><U+000C>%<U+0005><U+0008><U+0019><U+001E><U+0004><8B>b<80><U+0014>E<U+0002>B<92><U+000B>B<A4><U+0010>2<U+0014>8<U+0008><U+0018>I
                                                                                                                                                          ^
1:12:1: error: source file is not valid UTF-8
<90>!#<C4>R<80><U+000C><U+0019>!r$<U+0007><C8>H<U+0011>b<A8><A0><A8>@<C6><F0><U+0001>
^
1:12:4: error: source file is not valid UTF-8
<90>!#<C4>R<80><U+000C><U+0019>!r$<U+0007><C8>H<U+0011>b<A8><A0><A8>@<C6><F0><U+0001>
      ^
1:12:6: error: source file is not valid UTF-8
<90>!#<C4>R<80><U+000C><U+0019>!r$<U+0007><C8>H<U+0011>b<A8><A0><A8>@<C6><F0><U+0001>

.....~4000 lines....


>


)
```

The program was tested with `cl::sycl::cpu_selector` and that worked fine. It seems to be treating the compiled binary as text?

**Comment by rodburns** [comment]
Can you let me know how you are compiling the code and what device selector code you are using? 

Can you also double check your source code doesn't contain any characters (visible or not) that might be causing the compiler to think they are not standard UTF-8 characters?

**Comment by DuncanMcBain** [comment]
Hi @tom91136,
I think you're right, it does look like the GPU driver is treating the input as OpenCL C. This is a little strange as ComputeCpp actually adds `-x spir` to the `clBuildProgram` call to indicate that it's not an OpenCL C kernel. Have you tried a stable release from https://github.com/intel/compute-runtime/releases?

I'm also interested to know why `computecpp_info` thinks you only have one device if you have successfully run your code on the CPU. What does `clinfo` say?

**Comment by tom91136** [comment]
Sorry for the late reply.
I'm running this on University of Bristol's [Zoo](http://uob-hpc.github.io/zoo/) where we load OpenCL runtimes as modules. For cases where only the GPU is detected in `computecpp_info`, I've manually unloaded the [CPU runtime](https://software.intel.com/en-us/articles/opencl-drivers#cpu-section). 

The GPU driver used on the Iris Pro 580 is using `19.29.13530` which is a month old. For UHD 630, I'm running the latest stable version `19.32.13826`. 

The code in question is a port of [CloverLeaf](https://github.com/UoB-HPC/cloverleaf_kokkos) where we investigate performance portability and SYCL is one of our targets.

I was able to pinpoint the bug to `assert(...)` statements. Once those are removed, the program runs exactly like on the CPU. 
Looking a bit deeper, `assert` is implemented as a macro which delegates to: (`assert.h` - glibc on Linux):

```cpp
extern void __assert_fail (const char *__assertion, const char *__file,
			   unsigned int __line, const char *__function)
     __THROW __attribute__ ((__noreturn__));
```
Looking at the error, it does say at the beginning:

    error: undefined reference to `__assert_fail()'

I guess it's just unfortunate that the error looked like an assertion failure coming from the runtime.
To verify, calling other non-builtins such as `exit(1)` yields the same thing:

    error: undefined reference to `exit()'

I think this is basically resolved.
